server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.lucene.search.function;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.elasticsearch.ElasticsearchException;
+
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A query that uses a script to compute documents' scores.
+ */
+public class ScriptScoreQuery extends Query {
+    final Query subQuery;
+    final ScriptScoreFunction function;
+    private final Float minScore;
+
+    public ScriptScoreQuery(Query subQuery, ScriptScoreFunction function, Float minScore) {
+        this.subQuery = subQuery;
+        this.function = function;
+        this.minScore = minScore;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        Query rewritten = super.rewrite(reader);
+        if (rewritten != this) {
+            return rewritten;
```

Comment:
In most of the Lucene query implementations the call to `super.rewrite()` is done on the final return line (in this case line 66) so I wonder if its worth doing the same here for readability and consistency? I think the logic would end up being the same we would just be checking the internals of this  subclass first?